File: frontend/src/stores/category.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { Category } from '@/types'
import { categoryApi, learningApi, type CreateCategoryRequest, type UpdateCategoryRequest } from '@/api/client'

export const useCategoryStore = defineStore('category', () => {
  const categories = ref<Category[]>([])
  const currentCategory = ref<Category | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  const fetchCategories = async () => {
    loading.value = true
    error.value = null
    try {
      // Use public API for learning (no auth required)
      categories.value = await learningApi.getCategories()
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to fetch categories'
      throw err
    } finally {
      loading.value = false
    }
  }

  const fetchCategory = async (id: string) => {
    loading.value = true
    error.value = null
    try {
      const category = categories.value.find(c => c.id === id)
      if (category) {
        currentCategory.value = category
      }
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to fetch category'
      throw err
    } finally {
      loading.value = false
    }
  }

  const createCategory = async (data: CreateCategoryRequest) => {
    loading.value = true
    error.value = null
    try {
      const newCategory = await categoryApi.create(data)
      categories.value.push(newCategory)
      return newCategory
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to create category'
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateCategory = async (id: string, data: UpdateCategoryRequest) => {
    loading.value = true
    error.value = null
    try {
      const updatedCategory = await categoryApi.update(id, data)
      const index = categories.value.findIndex((c) => c.id === id)
      if (index !== -1) {
        categories.value[index] = updatedCategory
      }
      if (currentCategory.value?.id === id) {
        currentCategory.value = updatedCategory
      }
      return updatedCategory
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to update category'
      throw err
    } finally {
      loading.value = false
    }
  }

  const deleteCategory = async (id: string) => {
    loading.value = true
    error.value = null
    try {
      await categoryApi.delete(id)
      categories.value = categories.value.filter((c) => c.id !== id)
      if (currentCategory.value?.id === id) {
        currentCategory.value = null
      }
    } catch (err: any) {
      error.value = err.response?.data?.error || 'Failed to delete category'
      throw err
    } finally {
      loading.value = false
    }
  }

  return {
    categories,
    currentCategory,
    loading,
    error,
    fetchCategories,
    fetchCategory,
    createCategory,
    updateCategory,
    deleteCategory,
  }
})

File: backend/handlers/category.go
package handlers

import (
	"baby-words/database"
	"baby-words/models"
	"net/http"

	"github.com/gin-gonic/gin"
)

// GetCategories retrieves all categories
func GetCategories(c *gin.Context) {
	var categories []models.Category
	if err := database.DB.Order("sort_order ASC").Find(&categories).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch categories"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"data": categories})
}

// CreateCategoryRequest represents the request body for creating a category
type CreateCategoryRequest struct {
	ID        string `json:"id" binding:"required"`
	Title     string `json:"title" binding:"required"`
	Emoji     string `json:"emoji" binding:"required"`
	Gradient  string `json:"gradient" binding:"required"`
	SortOrder int    `json:"sort_order"`
}

// CreateCategory creates a new category
func CreateCategory(c *gin.Context) {
	var req CreateCategoryRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Check if category ID already exists
	var existingCategory models.Category
	if err := database.DB.First(&existingCategory, "id = ?", req.ID).Error; err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "Category ID already exists"})
		return
	}

	category := models.Category{
		ID:        req.ID,
		Title:     req.Title,
		Emoji:     req.Emoji,
		Gradient:  req.Gradient,
		SortOrder: req.SortOrder,
	}

	if err := database.DB.Create(&category).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create category"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"data": category})
}

// UpdateCategoryRequest represents the request body for updating a category
type UpdateCategoryRequest struct {
	Title     *string `json:"title"`
	Emoji     *string `json:"emoji"`
	Gradient  *string `json:"gradient"`
	SortOrder *int    `json:"sort_order"`
}

// UpdateCategory updates an existing category
func UpdateCategory(c *gin.Context) {
	id := c.Param("id")

	var category models.Category
	if err := database.DB.First(&category, "id = ?", id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Category not found"})
		return
	}

	var req UpdateCategoryRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Update fields if provided
	if req.Title != nil {
		category.Title = *req.Title
	}
	if req.Emoji != nil {
		category.Emoji = *req.Emoji
	}
	if req.Gradient != nil {
		category.Gradient = *req.Gradient
	}
	if req.SortOrder != nil {
		category.SortOrder = *req.SortOrder
	}

	if err := database.DB.Save(&category).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update category"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"data": category})
}

// DeleteCategory deletes a category (cascade deletes words)
func DeleteCategory(c *gin.Context) {
	id := c.Param("id")

	var category models.Category
	if err := database.DB.First(&category, "id = ?", id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Category not found"})
		return
	}

	if err := database.DB.Delete(&category).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete category"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Category deleted successfully"})
}
